File: reactor/r2dbc-mysql/r2dbc-mysql-be/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultInfo.java
package ua.mai.zyme.r2dbcmysql.exception;

import org.springframework.http.HttpStatus;

import java.text.MessageFormat;
import java.util.function.Supplier;

public interface FaultInfo extends Supplier<FaultException> {

    HttpStatus DEFAUILT_SERVICE_FAULT_ERROR_HTTPSTATUS =  HttpStatus.INTERNAL_SERVER_ERROR;  // 500

    String UNEXPECTED_ERROR_CODE = "ERR001";
    String UNEXPECTED_ERROR_TEMPLATE =  "Unexpected error: {0}";
    HttpStatus UNEXPECTED_ERROR_HTTPSTATUS =  DEFAUILT_SERVICE_FAULT_ERROR_HTTPSTATUS;

    String NOT_FOUND_CODE = "ERR002";
    String NOT_FOUND_TEMPLATE = "Resource {0} not found";
    HttpStatus NOT_FOUND_HTTPSTATUS =  HttpStatus.NOT_FOUND;

    String code();
    String messageTemplate();
    HttpStatus httpStatus();

    @Override
    default FaultException get() {
        return new FaultException(this);
    }

    static String createParamFor_NOT_FOUND(String resourceName, String resourceIdFieldName, String resourceIdValue) {
        return MessageFormat.format("{0} (for {1}={2})", resourceName, resourceIdFieldName, resourceIdValue);
    }

    static String createMessageFor_UNEXPECTED_ERROR(String errorMessage) {
        return MessageFormat.format(UNEXPECTED_ERROR_TEMPLATE, errorMessage);
    }

}

File: kotlin/kotlin-r2dbc-mysql/kotlin-r2dbc-mysql-be/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultInfo.kt
package ua.mai.zyme.r2dbcmysql.exception

import org.springframework.http.HttpStatus
import java.text.MessageFormat
import java.util.function.Supplier

interface FaultInfo : Supplier<FaultException> {

    companion object {
        val DEFAULT_SERVICE_FAULT_ERROR_HTTPSTATUS: HttpStatus = HttpStatus.INTERNAL_SERVER_ERROR  // 500

        const val UNEXPECTED_ERROR_CODE = "ERR001"
        const val UNEXPECTED_ERROR_TEMPLATE = "Unexpected error: {0}"
        val UNEXPECTED_ERROR_HTTPSTATUS: HttpStatus = DEFAULT_SERVICE_FAULT_ERROR_HTTPSTATUS

        const val NOT_FOUND_CODE = "ERR002"
        const val NOT_FOUND_TEMPLATE = "Resource {0} not found"
        val NOT_FOUND_HTTPSTATUS: HttpStatus = HttpStatus.NOT_FOUND

        fun createParamFor_NOT_FOUND(resourceName: String, resourceIdFieldName: String, resourceIdValue: String): String {
            return MessageFormat.format("{0} (for {1}={2})", resourceName, resourceIdFieldName, resourceIdValue)
        }

        fun createMessageFor_UNEXPECTED_ERROR(errorMessage: String): String {
            return MessageFormat.format(UNEXPECTED_ERROR_TEMPLATE, errorMessage)
        }
    }

    fun code(): String
    fun messageTemplate(): String
    fun httpStatus(): HttpStatus

    override fun get(): FaultException {
        return FaultException(this)
    }

}